Link back to metric and ping definitions in source control

Iinh opened the pull request:
Add a link back to where a metric/ping is originally defined in source control, as the second step of https://github.com/mozilla/glean-dictionary/issues/223. 

@wlach Could you review if this looks ok so far? Thanks!

wlach commented on probe_scraper/runner.py:
```diff
@@ -257,6 +257,38 @@ def load_moz_central_probes(
+    """
+    split_path = "".join(path).split("/")
+    for i, word in enumerate(split_path):
+        if len(word) == 40:  # i.e. a commit hash
```
I think it would be better to pass the commit hash (which we have) into this function and splitting on that -- that should be more reliable and simpler.

wlach commented on probe_scraper/runner.py:
```diff
@@ -280,16 +312,26 @@ def load_glean_metrics(cache_dir, out_dir, repositories_file, dry_run, glean_rep
         for commit_hash, paths in commits.items():
             metrics_files = [p for p in paths if p.endswith(GLEAN_METRICS_FILENAME)]
             pings_files = [p for p in paths if p.endswith(GLEAN_PINGS_FILENAME)]
-
```
Guessing you didn't mean to delete this line.

wlach commented on probe_scraper/runner.py:
```diff
@@ -280,16 +312,26 @@ def load_glean_metrics(cache_dir, out_dir, repositories_file, dry_run, glean_rep
-
             try:
                 config = {"allow_reserved": repo_name.startswith("glean")}
+                for d in repositories:
+                    if vars(d)["name"] == repo_name:
+                        repo_dict = d.to_dict()
```
I'm pretty sure this should work. Also, I think we should just use the name `repo` as opposed to `repo_dict`.

```suggestion
                repo = next(r for r in repositories if r.name == repo_name)
```

wlach reviewed (commented):
This looks pretty reasonable to me. It might be good to get someone like @fbertsch (who AFAIK wrote most of this glean-based code originally) to do the final review.

Iinh commented:
Okay I think I've addressed all of your review comments @wlach, please let me know if I've missed anything. And yes it'd be great if @fbertsch could do the final review for this PR. Thank you both!

wlach commented on probe_scraper/runner.py:
```diff
@@ -257,6 +257,38 @@ def load_moz_central_probes(
+    Output:
+    '530a5128526fcf133d57ec3d188590c600895fa1/glean-core/metrics.yaml'
+    """
+    split_path = "".join(path).split("/")
+    for i, word in enumerate(split_path):
+        if word == commit_hash:
+            file_path = "/".join(split_path[i:])
+    return file_path
```
I *think* you can replace this with this one-liner:

```suggestion
    return path.split(commit_hash)[-1][1:]
```

wlach reviewed (commented):
Can you give some example output from this? E.g. a metrics output for burnham.

Tagging @fbertsch for review as well.

fbertsch reviewed (changes requested):
We should push this field addition into the parsers themselves. The runner in general is overloaded, and we're trying to actively avoid adding new logic there.

It should be straightforward, since the parsers already receive a list of files. Note though that it is a _list_, not a single file - each Glean application can have more than one metrics or ping files. The only addition it needs is the url of the repository.

wlach commented:
> It should be straightforward, since the parsers already receive a list of files. Note though that it is a _list_, not a single file - each Glean application can have more than one metrics or ping files. The only addition it needs is the url of the repository.

Just to be clear, do you mean this change should happen inside [glean_parser](https://github.com/mozilla/glean_parser/)?

fbertsch commented:
> Just to be clear, do you mean this change should happen inside [glean_parser](https://github.com/mozilla/glean_parser/)?

I suppose it could. Either there or in `parsers/metrics.py`. We should check in with the client team to see if they have an opinion on including it in the glean-parser.

wlach commented:
> > Just to be clear, do you mean this change should happen inside [glean_parser](https://github.com/mozilla/glean_parser/)?
> 
> I suppose it could. Either there or in `parsers/metrics.py`. We should check in with the client team to see if they have an opinion on including it in the glean-parser.

@mdboom @badboy ^ ?

Iinh commented:
> @wlach Can you give some example output from this? E.g. a metrics output for burnham.

Here are some example outputs:
https://gist.github.com/Iinh/fa3b196aae8c375b695c560c28c351c3
https://gist.github.com/Iinh/e135dfc90e1b09873f6209c3af7e4e6b

badboy commented:
> > Just to be clear, do you mean this change should happen inside [glean_parser](https://github.com/mozilla/glean_parser/)?
> 
> I suppose it could. Either there or in `parsers/metrics.py`. We should check in with the client team to see if they have an opinion on including it in the glean-parser.

We already expose the line number, the fact that we don't also expose the file path within the repository seems like an oversight to me. `glean_parser` would indeed be the place where we have the data and can expose it.

badboy commented:
Actually, I would need to read the [PR on coverage data](https://github.com/mozilla/glean_parser/pull/272) again, but I think that tracks files already

wlach commented:
> > Actually, I would need to read the [PR on coverage data](https://github.com/mozilla/glean_parser/pull/272) again, but I think that tracks files already
> 
> It does, and it's a pretty simple change. That, of course, only tracks the file path within the repo -- glean_parser doesn't understand repos at all...

Perhaps a hybrid approach, where glean parser gives the path / line number and probe-scraper creates the final URL makes the most sense architecturally. Probably best if glean_parser doesn't *assume* that you're (e.g.) using GitHub.

badboy commented:
We're gonna land the linked PR, and I'm doing a release by tomorrow so we can use that data here.

wlach reviewed (changes requested):
@Iinh I think @fbertsch mentioned he'd prefer to see this logic (aside from what's now in glean_parser) moved into the parsers (`parsers/metrics.py` and `parsers/pings.py`, presumably). Could you give that a try?

fbertsch reviewed (commented):
@wlach @Iinh I'm looking at this again, and basically what we have now are 4 pieces:
- Repo URL
- Commit
- File path
- Line number

Since we added the latter two to the metric definitions, we should be able to materialize these URLs downstream from here; i.e. not include them directly in the probe info service.

wlach commented:
> Since we added the latter two to the metric definitions, we should be able to materialize these URLs downstream from here; i.e. not include them directly in the probe info service.

This is true, but I feel like it would be easiest / best to do this work inside probe-scraper. Otherwise the downstream consumer (the Glean Dictionary currently, but maybe other things later) has to put the information together, which I think is more work since it will involve gluing data together from a bunch of different data sources.

fbertsch commented:
> This is true, but I feel like it would be easiest / best to do this work inside probe-scraper. Otherwise the downstream consumer (the Glean Dictionary currently, but maybe other things later) has to put the information together, which I think is more work since it will involve gluing data together from a bunch of different data sources.

Got it. If you'd prefer it live here, I don't have any specific objections.

Iinh commented on probe_scraper/parsers/utils.py:
```diff
@@ -24,3 +24,15 @@ def get_major_version(version):
     return version.split(".")[0]
+
+
+def add_source_url(v, repo_url, commit_hash):
+    """Add source URL where metrics and pings are defined."""
+    file_path = v["defined_in"]["filepath"][
+        v["defined_in"]["filepath"].find(commit_hash) :  # noqa: E203
+    ]
+    line_number = v["defined_in"]["line"]
+    v["source_url"] = f"{repo_url}/blob/{file_path}#L{line_number}"
+    # the 'defined_in' structure is no longer needed
+    del v["defined_in"]
+    return v
```
@wlach I know how you feel about utils... but I wasn't sure where to put this function. Any advice?

wlach replied:
Since it's already there, I think this is fine.

wlach commented on probe_scraper/parsers/utils.py:
```diff
@@ -24,3 +24,15 @@ def get_major_version(version):
     return version.split(".")[0]
+
+
+def add_source_url(v, repo_url, commit_hash):
```
```suggestion
def get_source_url(defined_in, repo_url, commit_hash):
```

I think it's probably clearer if we make this function just return a source URL and pass in the `defined_in` part of Glean Parser's output. You can then do the various munging inline (it's a little bit of duplication, but not too much and I think worth it for the additional clarity).

fbertsch replied:
I might call this data structure `glean_definition`

fbertsch replied:
What Will is suggesting is to avoid side effects, which is a good pattern that really evolves from functional programming.

wlach reviewed (changes requested):
Looks good to me minus nits. @fbertsch wdyt?

fbertsch commented on probe_scraper/parsers/pings.py:
```diff
@@ -25,18 +27,21 @@ class GleanPingsParser:
-            errors,
-        )
+        pings = {
+            normalize_ping_name(ping_name): ping_data.serialize()
```
I would prefer the below addition live inside a separate function alongside this serialized call:
```suggestion
            normalize_ping_name(ping_name): generate_definition(ping_data, repo_url, commit_hash)
```

Iinh replied:
Hi @fbertsch could you help elaborate more on this? Since this function only takes `defined_in` and returns the source URL, I think what you recommended might not work. Maybe I'm still not understanding it quite right?

fbertsch replied:
No worries - what I'm proposing is a new function inside the `pings.py` file called `generate_definition`. That function would _call_ `get_source_url`, e.g.:

```
def generate_definition(ping_data, repo_url, commit_hash):
  serialized = ping_data.serialize()
  if repo_url and commit_hash:
    serialized["url"] = get_source_url(...)
  return serialized
```

fbertsch replied:
That way we only need to iterate over the values once, and it makes it obvious where any future changes to the definitions should live.

Iinh replied:
Ah this made a lot of sense! Thank you for explaining!

wlach commented on tests/test_metrics_parser.py:
```diff
@@ -21,3 +23,21 @@ def test_metrics_parser():
+
+    assert (
+        parsed_metrics["example.duration"]["source_url"]
+        == "test.com/foo/blob/tests/resources/metrics.yaml#L4"
```
I assume in production, the `repo_url` part includes a `https://` or similar? If it doesn't, maybe we should add it in `get_source_url`? If it does, maybe we should change our example `repo_url` to be more realistic.

Iinh replied:
Yes the `repo_url` from `repositories.yaml` include `https://`. I've updated the test URL to include `https://`

wlach reviewed (changes requested):
Code looks really good. I just have one last question / nit that I noticed when skimming the tests.

wlach reviewed (approved):
💯

fbertsch reviewed (approved):
LGTM, thanks for sticking with it @Iinh!
